src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs

```diff
@@ -12,30 +12,49 @@ public class SocketReceiver
         private readonly Socket _socket;
         private readonly SocketAsyncEventArgs _eventArgs = new SocketAsyncEventArgs();
         private readonly SocketAwaitable _awaitable;
+        private Memory<byte> _buffer;
 
         public SocketReceiver(Socket socket, PipeScheduler scheduler)
         {
             _socket = socket;
             _awaitable = new SocketAwaitable(scheduler);
-            _eventArgs.UserToken = _awaitable;
-            _eventArgs.Completed += (_, e) => ((SocketAwaitable)e.UserToken).Complete(e.BytesTransferred, e.SocketError);
+            _eventArgs.SetBuffer(Array.Empty<byte>(), 0, 0);
+            _eventArgs.UserToken = this;
+            _eventArgs.Completed += (_, e) => ((SocketReceiver)e.UserToken).OnReadable();
         }
 
         public SocketAwaitable ReceiveAsync(Memory<byte> buffer)
         {
-#if NETCOREAPP2_1
-            _eventArgs.SetBuffer(buffer);
-#else
-            var segment = buffer.GetArray();
-
-            _eventArgs.SetBuffer(segment.Array, segment.Offset, segment.Count);
-#endif
+            _buffer = buffer;
             if (!_socket.ReceiveAsync(_eventArgs))
             {
-                _awaitable.Complete(_eventArgs.BytesTransferred, _eventArgs.SocketError);
+                OnReadable();
             }
-
             return _awaitable;
         }
+
+        private void OnReadable()
```

Comment:
Instead of creating this OnReadable callback which is kind of hard to follow being callback base, wouldn't it be cleaner to do something along the lines of the following in SocketConnection.ProcessReceives?

```c#
                await _receiver.WaitForDataAsync();

                // Ensure we have some reasonable amount of buffer space
                var buffer = Input.GetMemory(MinAllocBufferSize);

                var bytesReceived = await _receiver.ReceiveAsync(buffer);
```

This has the additional benefit of delaying the allocation of the buffer until data is available which is a feature I'm pretty sure we want 😉

Reply:
Great suggestion! That will be a lot cleaner!